README.md
<<<<<<< HEAD
# phoenix_line
Flutter 企业级组件:line
=======
<!--
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages).

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages).
-->

phoenix 将做为企业级基础组件:Line组件,提供项目支持。

## Features

phoenix 将做为企业级基础组件:Line组件,提供项目支持。

## Getting started

phoenix 将做为企业级基础组件:Line组件,提供项目支持。

## Usage

phoenix 将做为企业级基础组件:Line组件,提供项目支持。

```dart
const like = 'sample';
```

## Additional information

phoenix 将做为企业级基础组件:Line组件,提供项目支持。
>>>>>>> 802a1d8 ([new] 新建)


CHANGELOG.md
<!--
 * @Author: lipeng 1162423147@qq.com
 * @Date: 2023-09-21 14:33:28
 * @LastEditors: lipeng 1162423147@qq.com
 * @LastEditTime: 2023-10-12 09:23:25
 * @FilePath: /phoenix_line/CHANGELOG.md
 * @Description: 这是默认设置,请设置`customMade`, 打开koroFileHeader查看配置 进行设置: https://github.com/OBKoro1/koro1FileHeader/wiki/%E9%85%8D%E7%BD%AE
-->
## 1.0.0

* [新增] phoenix 将做为企业级基础组件:Line组件,提供项目支持。

## 1.0.1

* [新增] example

test/phoenix_line_test.dart
/*
 * @Author: lipeng 1162423147@qq.com
 * @Date: 2023-09-21 14:33:28
 * @LastEditors: lipeng 1162423147@qq.com
 * @LastEditTime: 2023-09-21 14:46:57
 * @FilePath: /phoenix_line/test/phoenix_line_test.dart
 * @Description: 这是默认设置,请设置`customMade`, 打开koroFileHeader查看配置 进行设置: https://github.com/OBKoro1/koro1FileHeader/wiki/%E9%85%8D%E7%BD%AE
 */
import 'package:flutter_test/flutter_test.dart';

import 'package:phoenix_line/phoenix_line.dart';

void main() {
  test('adds one to input values', () {
    // final calculator = Calculator();
    // expect(calculator.addOne(2), 3);
    // expect(calculator.addOne(-7), -6);
    // expect(calculator.addOne(0), 1);
  });
}


lib/line.dart
/*
 * @Author: lipeng 1162423147@qq.com
 * @Date: 2022-04-29 17:06:50
 * @LastEditors: lipeng 1162423147@qq.com
 * @LastEditTime: 2023-10-12 09:21:39
 * @FilePath: /phoenix_line/lib/brn_line.dart
 * @Description: 这是默认设置,请设置`customMade`, 打开koroFileHeader查看配置 进行设置: https://github.com/OBKoro1/koro1FileHeader/wiki/%E9%85%8D%E7%BD%AE
 */
import 'package:flutter/material.dart';
import 'package:phoenix_base/phoenix.dart';

/// 用于分割页面组件元素的横向分割线
/// 支持自定义颜色、左右间距、和高度
///
/// 系统提供的[Divider]组件，当设置的高度过大时，会出现中间一条线，
///
/// ```dart
///   Line()
///
///   Line(
///      leftInset: 20,
///   )
///
/// ```
///
/// 相关分割线如下:
///
class Line extends StatelessWidget {
  /// 分割线的或者分割条的颜色 可以通过[BrnThemeConfigurator]配置默认颜色
  final Color? color;

  /// 分割线的或者分割条的高度 默认0.5
  final double height;

  /// 左边缩进距离
  final double leftInset;

  /// 右边缩进距离
  final double rightInset;

  const Line({
    Key? key,
    this.color,
    this.height = 0.5,
    this.leftInset = 0,
    this.rightInset = 0,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Padding(
      padding: EdgeInsets.only(left: leftInset, right: rightInset),
      child: Divider(
        thickness: height,
        height: height,
        color: color ??
            BaseThemeConfig.instance.getConfig().commonConfig.dividerColorBase,
      ),
    );
  }
}


lib/dashed_line.dart
import 'package:flutter/material.dart';
import 'package:phoenix_base/phoenix.dart';

/// 描述: 自定义界面元素中的虚线分割线
///
/// 适用场景
/// 1、界面元素中的分割线
///
/// 2、分割线样式的装饰

/// 分割线所在位置
enum DashedLinePosition {
  /// 头部
  trailing,

  /// 尾部
  leading,
}

/// 默认虚线方向
const Axis _normalAxis = Axis.horizontal;

/// 默认虚线长度
const double _normalDashedLength = 8;

/// 默认虚线厚度
const double _normalDashedThickness = 1;

/// 默认间距
const double _normalDashedSpacing = 4;

/// 默认位置，头部
const DashedLinePosition _normalPosition = DashedLinePosition.leading;

/// 虚线分割线
class DashedLine extends StatelessWidget {
  /// 虚线方向，默认值[_normalAxis]
  final Axis axis;

  /// 虚线长度，默认值[_normalDashedLength]
  final double dashedLength;

  /// 虚线厚度，默认值[_normalDashedThickness]
  final double dashedThickness;

  /// 虚线间距，默认值[_normalDashedSpacing]
  final double dashedSpacing;

  /// 颜色，默认值[_normalColor]
  final Color? color;

  /// 虚线的Widget
  final Widget contentWidget;

  /// 距离边缘的位置（偏移量），默认值为0
  final double dashedOffset;

  /// 分割线所在位置，默认值[_normalPosition]
  final DashedLinePosition position;

  const DashedLine({
    super.key,
    required this.contentWidget,
    this.axis = _normalAxis,
    this.dashedLength = _normalDashedLength,
    this.dashedThickness = _normalDashedThickness,
    this.dashedSpacing = _normalDashedSpacing,
    this.color,
    this.dashedOffset = 0.0,
    this.position = _normalPosition,
  });

  @override
  Widget build(BuildContext context) {
    return CustomPaint(
      painter: DashedPainter(
          axis: axis,
          dashedLength: dashedLength,
          dashedThickness: dashedThickness,
          dashedSpacing: dashedSpacing,
          color: color,
          dashedOffset: dashedOffset,
          position: position),
      child: contentWidget,
    );
  }
}

class DashedPainter extends CustomPainter {
  /// 虚线方向
  final Axis axis;

  /// 虚线长度
  final double dashedLength;

  /// 虚线厚度
  final double dashedThickness;

  /// 虚线间距
  final double dashedSpacing;

  /// 颜色
  final Color? color;

  /// 距离边缘的位置
  final double? dashedOffset;

  /// 分割线所在位置
  final DashedLinePosition? position;

  DashedPainter({
    this.axis = _normalAxis,
    this.dashedLength = _normalDashedLength,
    this.dashedThickness = _normalDashedThickness,
    this.dashedSpacing = _normalDashedSpacing,
    this.color,
    this.dashedOffset = 0.0,
    this.position = _normalPosition,
  });

  @override
  void paint(Canvas canvas, Size size) {
    var paint = Paint() // 创建一个画笔并配置其属性
      ..strokeWidth = dashedThickness // 画笔的宽度
      ..isAntiAlias = true // 是否抗锯齿
      ..color = color ??
          BaseThemeConfig.instance
              .getConfig()
              .commonConfig
              .dividerColorBase; // 画笔颜色

    var maxWidth = size.width; // size获取到宽度
    var maxHeight = size.height; // size获取到宽度
    if (axis == Axis.horizontal) {
      // 水平方向
      double startX = 0;
      final space = (dashedSpacing + dashedLength);
      double height = 0;
      if (position == DashedLinePosition.leading) {
        // 头部
        height = dashedOffset! + dashedThickness / 2;
      } else {
        // 尾部
        height = size.height - dashedOffset! - dashedThickness / 2;
      }
      while (startX < maxWidth) {
        if ((maxWidth - startX) < dashedLength) {
          canvas.drawLine(Offset(startX, height),
              Offset(startX + (maxWidth - startX), height), paint);
        } else {
          canvas.drawLine(Offset(startX, height),
              Offset(startX + dashedLength, height), paint);
        }
        startX += space;
      }
    } else {
      // 垂直方向
      double startY = 0;
      final space = (dashedSpacing + dashedLength);
      double width = 0;
      if (position == DashedLinePosition.leading) {
        // 头部
        width = dashedOffset! + dashedThickness / 2;
      } else {
        // 尾部
        width = size.width - dashedOffset! - dashedThickness / 2;
      }
      while (startY < maxHeight) {
        if ((maxHeight - startY) < dashedLength) {
          canvas.drawLine(Offset(width, startY),
              Offset(width, startY + (maxHeight - startY)), paint);
        } else {
          canvas.drawLine(Offset(width, startY),
              Offset(width, startY + dashedLength), paint);
        }
        startY += space;
      }
    }
  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) => true;
}


example/lib/main.dart
import 'package:flutter/material.dart';

import 'dashed_line_example.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // TRY THIS: Try running your application with "flutter run". You'll see
        // the application has a blue toolbar. Then, without quitting the app,
        // try changing the seedColor in the colorScheme below to Colors.green
        // and then invoke "hot reload" (save your changes or press the "hot
        // reload" button in a Flutter-supported IDE, or press "r" if you used
        // the command line to start the app).
        //
        // Notice that the counter didn't reset back to zero; the application
        // state is not lost during the reload. To reset the state, use hot
        // restart instead.
        //
        // This works for code too, not just values: Most code changes can be
        // tested with just a hot reload.
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: DashedLineExample(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      // This call to setState tells the Flutter framework that something has
      // changed in this State, which causes it to rerun the build method below
      // so that the display can reflect the updated values. If we changed
      // _counter without calling setState(), then the build method would not be
      // called again, and so nothing would appear to happen.
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return Scaffold(
      appBar: AppBar(
        // TRY THIS: Try changing the color here to a specific color (to
        // Colors.amber, perhaps?) and trigger a hot reload to see the AppBar
        // change color while the other colors stay the same.
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(widget.title),
      ),
      body: Center(
        // Center is a layout widget. It takes a single child and positions it
        // in the middle of the parent.
        child: Column(
          // Column is also a layout widget. It takes a list of children and
          // arranges them vertically. By default, it sizes itself to fit its
          // children horizontally, and tries to be as tall as its parent.
          //
          // Column has various properties to control how it sizes itself and
          // how it positions its children. Here we use mainAxisAlignment to
          // center the children vertically; the main axis here is the vertical
          // axis because Columns are vertical (the cross axis would be
          // horizontal).
          //
          // TRY THIS: Invoke "debug painting" (choose the "Toggle Debug Paint"
          // action in the IDE, or press "p" in the console), to see the
          // wireframe for each widget.
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            const Text(
              'You have pushed the button this many times:',
            ),
            Text(
              '$_counter',
              style: Theme.of(context).textTheme.headlineMedium,
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: const Icon(Icons.add),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}


example/lib/dashed_line_example.dart
/*
 * @Author: lipeng 1162423147@qq.com
 * @Date: 2022-04-29 17:06:50
 * @LastEditors: lipeng 1162423147@qq.com
 * @LastEditTime: 2023-10-12 09:21:57
 * @FilePath: /phoenix_line/example/lib/dashed_line_example.dart
 * @Description: 这是默认设置,请设置`customMade`, 打开koroFileHeader查看配置 进行设置: https://github.com/OBKoro1/koro1FileHeader/wiki/%E9%85%8D%E7%BD%AE
 */
import 'package:flutter/material.dart';
import 'package:phoenix_line/phoenix_line.dart';

class DashedLineExample extends StatefulWidget {
  @override
  _DashedLineExampleState createState() => _DashedLineExampleState();
}

class _DashedLineExampleState extends State<DashedLineExample> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.white,
      appBar: AppBar(
        title: const Text('虚线分割线'),
      ),
      body: SingleChildScrollView(
        child: Column(
          mainAxisSize: MainAxisSize.min,
          crossAxisAlignment: CrossAxisAlignment.start,
          children: <Widget>[
            const Text(
              '分割线的空间是由内部内容撑开',
              style: TextStyle(
                color: Color(0xFF222222),
                fontSize: 18,
              ),
            ),
            // 分割线的空间是由内部内容撑开
            DashedLine(
              dashedLength: 20,
              dashedThickness: 5,
              axis: Axis.vertical,
              color: Colors.red,
              dashedOffset: 20,
              position: DashedLinePosition.leading,
              contentWidget: Container(
                margin: const EdgeInsets.only(
                    left: 60, right: 20, top: 10, bottom: 10),
                child: const Text(
                    "穿插介绍、公司模式一句话C端服务承诺介绍、价值穿插介绍、公司模式一句话C端服务承诺介绍、价值穿插介绍、公司模式一句话C端服务承诺介绍、价值"),
              ),
            ),
            const Text(
              '分割线的空间是由内部容器设定',
              style: TextStyle(
                color: Color(0xFF222222),
                fontSize: 18,
              ),
            ),
            // 分割线的空间是由内部容器设定
            DashedLine(
              dashedLength: 10,
              dashedThickness: 3,
              axis: Axis.horizontal,
              color: Colors.green,
              dashedOffset: 20,
              position: DashedLinePosition.leading,
              contentWidget: Container(
                width: 200,
                height: 100,
              ),
            ),
            const Text(
              '分割线的空间由外部设定',
              style: TextStyle(
                color: Color(0xFF222222),
                fontSize: 18,
              ),
            ),
            // 分割线的空间由外部设定
            Container(
              height: 50,
              width: 300,
              padding: const EdgeInsets.all(5),
              color: Colors.red,
              child: DashedLine(
                axis: Axis.horizontal,
                dashedOffset: 10,
                contentWidget: Container(
                  width: 200,
                  height: 100,
                ),
              ),
            ),
          ],
        ),
      ),
    );
  }
}
